clients/amqp-common/src/main/java/org/eclipse/hono/client/amqp/AbstractRequestResponseServiceClient.java

```diff
@@ -299,6 +301,44 @@ protected final void addToCache(final Object key, final R response) {
         }
     }
 
+    /**
+     * Removes a response from the cache.
+     * <p>
+     * If no cache is configured then this method does nothing.
+     *
+     * @param key The key to be removed.
+     * @throws NullPointerException if key is {@code null}.
+     */
+    protected final void removeFromCache(final Object key) {
+        if (isCachingEnabled()) {
+            Objects.requireNonNull(key);
+            responseCache.invalidate(key);
+        }
+    }
+
+    /**
+     * Removes responses from the cache where the keys match the given predicate.
+     * <p>
+     * If no cache is configured then this method does nothing.
+     *
+     * @param keyPredicate The predicate to filter the affected keys.
+     * @throws NullPointerException if keyPredicate is {@code null}.
+     */
+    protected final void removeFromCacheByPattern(final Predicate<Object> keyPredicate) {
+        if (isCachingEnabled()) {
+            Objects.requireNonNull(keyPredicate);
+
+            final var matchingKeys = responseCache.asMap()
+                    .keySet()
+                    .stream()
+                    .filter(keyPredicate)
+                    .collect(Collectors.toSet());
```

Comment:
I am a little concerned about the runtime behavior of this when the cache has a lot of entries. Apart from blocking the event loop runtime seems to scale linearly with the cache size. This is probably not an issue with the tenant cache but for devices and/or credentials with dozens of thousands of entries I am not so sure ...

Reply:
@calohmn @kaniyan Do you have any ideas on this?

Reply:
Here is an idea that also involves additional data structures (but I guess is a bit different than envisioned above).
The following description is for the device registration client:

We could keep an extra cache with received device-change notifications, with an eviction timespan that is equivalent to the eviction timespan of the device registration cache.
This device-change cache could have tenant+device id as key, and the notification value (including timestamp) as value.

When a device registration client `assertRegistration` method gets called, there is first the check for a corresponding entry in the device registration cache. 
If a value is found there, 2 extra lookups on the device-change cache are done (with keys "tenant+device id" and "tenant+gateway id"). If there is a cache hit on that, cache timestamps are to be compared: If the device-change cache entry timestamp is newer than the device registration cache entry timestamp, a new device registry request has to be done to get the current device registration data and then update the device registration cache entry.

It would be good to be able to delete the device-change cache entry once a corresponding device registration cache entry has been updated.
However, this gets more complicated when there was a device-change notification regarding a gateway entry, because this may affect multiple device registration cache entries. Maybe if we add some info in the notification message about whether the device is a gateway, we can apply some optimizations here.
To react on *tenant*-change notifications with this approach, we could possible store entries in the device-change cache here as well - with the device id part of the cache key being null.

For the credentials data cache, we could think of a similar approach.

Reply:
Thank you @calohmn for your proposal! I'll have a closer look on it later on.

Two ideas of mine that are independent of this:

1. It might be possible to achieve more efficient scaling with the existing solution by means of sorting.
2. Alternatively, I wonder if it wouldn't be sufficient to take the operation off the event loop (perhaps combined with a parallel stream). That might be sufficient, assuming that these notifications are relatively rare events that are also not particularly time-critical. WDYT?

Reply:
I wonder if we can simply assume that we do not need to invalidate all entries belonging to a tenant from the registration and credentials caches. In [this comment](https://github.com/eclipse/hono/issues/2984#issuecomment-987703829) @calohmn argued that we can assume that clients would not use values from the registration nor credentials caches if the tenant does not exist or is disabled. If that is true then I guess we can simply wait for the entries to be evicted from the cache instead of actively removing them when a tenant gets deleted, or am I missing something?

Reply:
> I wonder if we can simply assume that we do not need to invalidate all entries belonging to a tenant from the registration and credentials caches.

My understanding of the code is that negative results are also cached. This means that a tenant would not be available after creation until the cache expires.
In addition, there is the case that all devices of a tenant are deleted. Then the tenant still exists and is enabled. For this case, there is a separate notification, where we still have the problem to find the affected devices and credentials only by means of the tenant ID.

Reply:
> > I wonder if we can simply assume that we do not need to invalidate all entries belonging to a tenant from the registration and credentials caches.

> My understanding of the code is that negative results are also cached. This means that a tenant would not be available after creation until the cache expires.

But the tenant creation notification would invalidate the cache entry here, wouldn't it? Or which scenario were you thinking of here?

> In addition, there is the case that all devices of a tenant are deleted. 

I guess in practise, this operation would most probably be followed by a deletion of the tenant.  So if we assume that the tenant client is invoked first, the device/credentials client invocations wouldn't happen here anyway.

Reply:
> My understanding of the code is that negative results are also cached.

That shouldn't be the case. What makes you believe so?

Reply:
Nvermind. I just found the place that prevents negative results from being cached (for the record: `ProtonBasedDeviceRegistrationClient.getResult()` sets the `cacheDirective` property of `RegistrationResult` to `null` and then `AbstractRequestResponseServiceClient.addToCache()` filters for status codes).

Reply:
I will change the PR so that notifications about the creation of tenants, devices, and credentials are not processed because negative results are not cached.
Also, I will no longer clean up the Registration Assertions and Credentials caches when deleting the tenant. This way the cleanup will take place less often.

Reply:
> > In addition, there is the case that all devices of a tenant are deleted.
> 
> I guess in practise, this operation would most probably be followed by a deletion of the tenant. 

@sophokles73 Do you agree that we can assume that users will delete the tenant shortly after they invoked the operation that removes all devices and therefore skipping the cleanup of registration and credentials caches?

Reply:
I tend to agree. The only use case that comes to mind where it wouldn't is that you might want to re-use a (test-)tenant for a new series of tests but want to start with a fresh set of devices, thus removing all registered devices before running the tests. However, I guess that this notification will be sent very seldomly in comparison to the device/credentials updated/removed notifications ...

Reply:
@sophokles73  Makes sense. For the sake of consistency and because we already agreed to use a worker thread I tent to clean up registration and credentials caches on that notification.